File: datastruct/elementary/sequence/src/bralist.c
/*
 * bralist.c
 * Copyright (C) 2012 Liu Xinyu (liuxinyu95@gmail.com)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * Binary random access list based on [1]
 * [1]. Chris Okasaki. ``Purely Functional Random-Access Lists''. 
 * Functional Programming Languages and Computer Architecutre, June 1995, pages 86-95.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h> /* for memcpy */
#include <assert.h> /* for assertion */

#define BIG_RAND() rand() % 1000

#define M sizeof(int)*8      /* Limit the length of the list to 2^(M+1) - 1 */
typedef int Key;  /* Alternatively,  C++ template can be used */

struct List {
  int n;
  Key* tree[M];
};

struct List empty() {
  int i;
  struct List a;
  a.n = 0;
  for(i = 0; i < M; ++i)
    a.tree[i] = NULL;
  return a;
}

void release(struct List a) {
  int i;
  for(i = 0; i < M; ++i)
    free(a.tree[i]);   /* free(NULL) has no side effect */
}

int size(struct List a) {
  return a.n;
}

/* returns number of bits of n */
int nbits(int n) {
  int i=0;
  while(n >>= 1)
    ++i;
  return i;
}

/*
 * Aggregate performance analysis.
 */
struct List insert(struct List a, Key x) {
  int i, j, sz;
  Key* xs;
  i = nbits( (a.n+1) ^ a.n ); /* xor helps locate the bit flip to 1 r*/
  xs = a.tree[i] = (Key*)malloc(sizeof(Key)*(1<<i));
  for(j=0, *xs++ = x, sz = 1; j<i; ++j, sz <<= 1) {
    memcpy((void*)xs, (void*)a.tree[j], sizeof(Key)*(sz));
    xs += sz;
    free(a.tree[j]);
    a.tree[j] = NULL;
  }
  ++a.n;
  return a;
}

/* 
 * O(\lg N) random access 
 * index out of bound error is omit here
 * suppose i < size(a)
 */
Key get(struct List a, int i) {
  int j, sz;
  for(j = 0, sz = 1; j < M; ++j, sz <<= 1)
    if(a.tree[j]) {
      if(i < sz)
	break;
      i -= sz;
    }
  return a.tree[j][i];
}

void print_array(Key* xs, int n) {
  int i;
  for(i=0; i<n; ++i)
    printf( i == n -1 ? "%d\n" : "%d, ", xs[i]);
}

void print_list(struct List a) {
  int i;
  for(i=0; i<size(a); ++i)
    printf( i == size(a) -1 ? "%d\n" : "%d, ", get(a, i));
}

void print_trees(struct List a) {
  int i, j, m, n;
  printf("size: %d:\n", a.n);
  m = nbits(a.n);
  for(i=0; i<=m; ++i) {
    printf("tree[%d]:", i);
    if(a.tree[i]) {
      n = 1<<i;
      for(j=0; j<n; ++j)
	printf(j==n-1 ? "%d\n" : "%d, ", a.tree[i][j]);
    }
    else
      printf("NIL\n");
  }
  printf("\n");
}

void assert_eq(Key* xs, struct List ys, int n) {
  int i;
  for(i = 0; i<n; ++i)
    assert(xs[i] == get(ys, i));
}

void test_insert() {
  int i, n, m=1000;
  Key* xs;
  struct List ys;
  while(m--){
    n = BIG_RAND();
    xs = (Key*)malloc(sizeof(Key)*n);
    for(i = 0; i<n; ++i)
      xs[i] = BIG_RAND();
    ys = empty();
    for(i=n-1; i>=0; --i)
      ys = insert(ys, xs[i]);
    assert_eq(xs, ys, n);
    release(ys);
    free(xs);
  }
}

int main(int argc, char** argv) {
  test_insert();
  return 0;
}

File: datastruct/elementary/sequence/src/BRAList.hs
{-
    BRAList.hs, Binary Random Access List without numeric representation.
    Copyright (C) 2011, Liu Xinyu (liuxinyu95@gmail.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-}

module BRAList where

import Test.QuickCheck

-- Based on Chris Okasaki's work in [1]

-- Binary tree representation

data Tree a = Leaf a
            | Node Int (Tree a) (Tree a) -- size, left, right
              deriving (Show)

-- The random access list is represented as a forest of binary trees.
type BRAList a = [Tree a]

size (Leaf _) = 1
size (Node sz _ _) = sz

link t1 t2 = Node (size t1 + size t2) t1 t2

insert x = insertTree (Leaf x) where
  insertTree t [] = [t]
  insertTree t (t':ts) = if size t < size t' then  t:t':ts
                         else insertTree (link t t') ts

-- Assert the list isn't empty
extract ((Leaf x):ts) = (x, ts)
extract ((Node _ t1 t2):ts) = extract (t1:t2:ts)

head' = fst . extract

tail' = snd . extract

getAt _ [] = Nothing
getAt i (t:ts) | i < 0 = Nothing
               | i < size t = lookupTree i t
               | otherwise = getAt (i - size t) ts
  where
    lookupTree 0 (Leaf x) = Just x
    lookupTree i (Node sz t1 t2) | i < sz `div` 2 = lookupTree i t1
                                 | otherwise = lookupTree (i - sz `div` 2) t2

setAt (t:ts) i x = if i < size t then (updateTree t i x):ts
                   else t:setAt ts (i-size t) x
  where
    updateTree (Leaf _) 0 x = Leaf x
    updateTree (Node sz t1 t2) i x =
        if i < sz `div` 2 then Node sz (updateTree t1 i x) t2
        else Node sz t1 (updateTree t2 (i - sz `div` 2) x)

fromList = foldr insert []

toList [] = []
toList (t:ts) = (treeToList t) ++ toList ts where
    treeToList (Leaf x) = [x]
    treeToList (Node _ t1 t2) = (treeToList t1) ++ treeToList t2

-- testing
prop_insert :: [Int] -> Bool
prop_insert xs = xs == (toList $ fromList xs)

prop_head :: [Int] -> Property
prop_head xs = not (null xs) ==> xs == (rebuild $ fromList xs) where
    rebuild [] = []
    rebuild ts = head' ts : (rebuild $ tail' ts)

prop_lookup :: [Int] -> Int -> Bool
prop_lookup xs i = case getAt i (fromList xs) of
  Nothing -> i < 0 || i >= length xs
  Just x -> x == (xs !! i)

prop_update :: [Int] -> Int -> Int -> Property
prop_update xs i y = (0 <=i && i< length xs) ==> toList (setAt (fromList xs) i y) == xs' where
    xs' = as ++ [y] ++ bs
    (as, (_:bs)) = splitAt i xs

testAll = do
  quickCheck prop_insert
  quickCheck prop_head
  quickCheck prop_lookup
  quickCheck prop_update

-- Reference
-- [1]. Chris Okasaki. ``Purely Functional Random-Access Lists''. Functional Programming Languages and Computer Architecutre, June 1995, pages 86-95.
